frontends/p4/duplicateActionControlPlaneNameCheck.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2024 Cisco Systems, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "duplicateActionControlPlaneNameCheck.h"
+
+#include "lib/log.h"
+
+namespace P4 {
+
+cstring DuplicateActionControlPlaneNameCheck::getName(const IR::IDeclaration *decl) {
+    return decl->getName();
+}
+
+void DuplicateActionControlPlaneNameCheck::checkForDuplicateName(cstring name,
+                                                                 const IR::Node *node) {
+    bool foundDuplicate = false;
+    auto *otherNode = node;
+    auto [it, inserted] = actions.insert(std::pair(name, node));
+    if (!inserted) {
+        foundDuplicate = true;
+        otherNode = (*it).second;
+    }
+    if (foundDuplicate) {
+        error(ErrorType::ERR_DUPLICATE, "%1%: conflicting control plane name: %2%", node,
+              otherNode);
+    }
+}
+
+const IR::Node *DuplicateActionControlPlaneNameCheck::postorder(IR::P4Action *action) {
+    bool topLevel = findContext<IR::P4Control>() == nullptr;
```

Comment:
```
bool topLevel = !stack.empty();
```
has some chances to be a tiny bit more efficient, I believe.
(this requires the previously suggested change for not pushing parsers onto the stack).

Reply:
Changed to `bool topLevel = stack.empty();` in commit 57